tests/ui_tests/reporting/report_master_diff.py

```diff
@@ -0,0 +1,191 @@
+):
+    doc = dominate.document(title=test_name)
+    master_screens = sorted(master_screens_path.iterdir())
+    current_screens = sorted(current_screens_path.iterdir())
+
+    with doc:
+        h1(test_name)
+        p("This UI test differs from master.", style="color: grey; font-weight: bold;")
+        with table():
+            with tr():
+                td("Master:")
+                td(master_hash, style="color: red;")
+            with tr():
+                td("Current:")
+                td(current_hash, style="color: green;")
+        hr()
+
+        with table(border=1, width=600):
+            with tr():
+                th("Master")
+                th("Current branch")
+
+            html.diff_table(master_screens, current_screens)
+
+    return html.write(REPORTS_PATH / "diff", doc, test_name + ".html")
+
+
+def index():
+    removed = list((REPORTS_PATH / "removed").iterdir())
+    added = list((REPORTS_PATH / "added").iterdir())
+    diff = list((REPORTS_PATH / "diff").iterdir())
+
+    title = "UI changes from master"
+    doc = dominate.document(title=title)
+
+    with doc:
+        h1("UI changes from master")
+        hr()
+
+        h2("Removed:", style="color: red;")
+        i("UI fixtures that have been removed:")
+        html.report_links(removed, REPORTS_PATH)
+        br()
+        hr()
+
+        h2("Added:", style="color: green;")
+        i("UI fixtures that have been added:")
+        html.report_links(added, REPORTS_PATH)
+        br()
+        hr()
+
+        h2("Differs:", style="color: grey;")
+        i("UI fixtures that have been modified:")
+        html.report_links(diff, REPORTS_PATH)
+
+    return html.write(REPORTS_PATH, doc, "index.html")
+
+
+def create_dirs():
+    # delete the reports dir to clear previous entries and create folders
+    shutil.rmtree(REPORTS_PATH, ignore_errors=True)
+    REPORTS_PATH.mkdir()
+    (REPORTS_PATH / "removed").mkdir()
+    (REPORTS_PATH / "added").mkdir()
+    (REPORTS_PATH / "diff").mkdir()
+
+
+def create_reports():
+    removed_tests, added_tests, diff_tests = get_diff()
+
+    with tempfile.TemporaryDirectory(prefix="trezor-records-") as temp_dir:
+        temp_dir = Path(temp_dir)
+
+        for test_name, test_hash in removed_tests.items():
+            download.fetch_recorded(test_hash, temp_dir)
+            removed(temp_dir, test_name)
+
+        for test_name, test_hash in added_tests.items():
+            path = RECORDED_SCREENS_PATH / test_name / "actual"
+            if not path.exists():
+                raise RuntimeError("Folder does not exist, has it been recorded?", path)
+            added(path, test_name)
+
+        for test_name, test_hashes in diff_tests.items():
+            master_screens = temp_dir
+            download.fetch_recorded(test_hashes["master"], master_screens)
+
+            current_screens = (
+                RECORDED_SCREENS_PATH / test_name / "actual"
+            )  # TODO: use hash
```

Comment:
This was a note to myself, which I forgot to remove. I have realized that we currently name the folders in `ui_tests/screens` using the test name. But we should call it with the test name and the hash to make sure we include the correct revision of the test.

For example, if you have not run `ui=test` for some time, but you have _some_ files in the  screens directory we will use those files for the report, which is incorrect. On the other hand this should not happen on CI, so I have decided to omit it for the time being. What do you think?

Reply:
you could also recalculate the hash from the saved screens

i'm not sure if it's worth having the hash in the directory name, because this (report_master_diff) would be the only thing using them, right?

i think it's OK to leave it as it is now

Reply:
Yeah that is true. I agree let's leave it like that for now.